### ISSUE TITLE
Project build not works

### ISSUE THREAD

**Issue by djKooks** [cloudflare/wrangler2#254]
Hello,
I've tried following build process, but it just keep showing logs with build command
```
$ npm run build            

> wrangler-root@0.0.0 build /Users/kination/workspace/oss/wrangler2
> npm run build --workspace=wrangler

> wrangler-root@0.0.0 build /Users/kination/workspace/oss/wrangler2
> npm run build --workspace=wrangler

> wrangler-root@0.0.0 build /Users/kination/workspace/oss/wrangler2
> npm run build --workspace=wrangler

> wrangler-root@0.0.0 build /Users/kination/workspace/oss/wrangler2
> npm run build --workspace=wrangler
...
...
```
Am I doing correctly, or missing something?

**Comment by petebacondarwin** [comment]
Hi @djKooks - thanks for getting in touch.

Can you provide the following to help us:
- your OS
- your node and npm versions

Also can you clean out node_modules from both the top level and under packages/wrangler.

**Comment by djKooks** [comment]
@petebacondarwin thanks for follow-up 🙏 
- OS: Mac M1 (Big Sur)
- Node / NPM: `v14.16.1` / `6.14.12`

I've follow-up the process in this document (https://github.com/cloudflare/wrangler2/blob/main/CONTRIBUTING.md), so there were no `node_modules` in `package/wrangler`.

**Comment by petebacondarwin** [comment]
Please try node version 16 (the latest LTS version), which will also result in npm version 8

**Comment by djKooks** [comment]
@petebacondarwin thanks!
It seems working after bumping up node.

**Comment by petebacondarwin** [comment]
Great - glad it is working. The CONTRIBUTING doc does mention using the LTS version. But perhaps we should make that more clear? Closing this issue as you are now unblocked.

**Comment by djKooks** [comment]
@petebacondarwin thanks!
Maybe it will be better to make it more clear, unless if you're thinking to make this project to always base on newest node version.

**Comment by petebacondarwin** [comment]
I think we should always be encouraging developers of the wrangler tool to be using the latest LTS version of node.
The version that wrangler runs on could be more relaxed...

**Comment by mrbbot** [comment]
The requirement for `>= 16.7.0` comes from Miniflare, which makes extensive use of new Web Standards added in recent versions of Node.js (Web Streams, `crypto`, `Blob`, `EventTarget`, etc). The most recent thing Miniflare uses is the [`stream/consumers`](https://nodejs.org/api/webstreams.html#utility-consumers) module, which was added in 16.7.0.

**Comment by djKooks** [comment]
@petebacondarwin @mrbbot thanks for comment. Looks very reasonable 👍

**Comment by petebacondarwin** [comment]
Indeed we even have

```
  "engines": {
    "node": ">=16.0.0"
  },
``` 

in our top level package.json.

We should update this to: `>= 16.7.0`. (And perhaps enforce it in a postinstall script?

**Comment by djKooks** [comment]
@petebacondarwin looks right. How do you think of following change?
https://github.com/cloudflare/wrangler2/pull/271

**Comment by djKooks** [comment]
Also one more question.
When I try to run `wrangler dev index.js`, it shows as following after account creation.
```
⬣ Listening at http://localhost:8787
remote worker: Error: Failed to fetch /accounts/0d03575700684a90f4af1b90e87be426/workers/subdomain/edge-preview - 403: Forbidden
    at fetchInternal (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:109478:11)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async fetchResult (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:109509:16)
    at async sessionToken (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:111903:28)
    at async previewToken (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:111921:53)
    at async createWorker (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:111945:17)
    at async start (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:113032:16)
```
Is there something more need to be done?
(or link of guide document is also okay 🙏 )

**Comment by petebacondarwin** [comment]
Not sure about that error... @threepointone ??

**Comment by threepointone** [comment]
Could you try with the latest beta? I think this was a problem with 0.0.8. Install `wrangler@beta` and see if this still occurs?

**Comment by threepointone** [comment]
I think I figured this out, are you using durable objects by any chance?

**Comment by djKooks** [comment]
@threepointone you mean in `index.js`? I've just followed up the process in quickstart
```
echo "export default { fetch() { return new Response('hello world') } }" > index.js
```

**Comment by djKooks** [comment]
@threepointone @petebacondarwin okay this is the one I can see in the newest commit version
```
wrangler-test wrangler dev index.js
⬣ Listening at http://localhost:8787
remote worker: Error: Failed to fetch /accounts/0d03575700684a90f4af1b90e87be426/workers/subdomain/edge-preview - 403: Forbidden
    at fetchInternal (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:109478:11)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async fetchResult (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:109509:16)
    at async sessionToken (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:111903:28)
    at async previewToken (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:111921:53)
    at async createWorker (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:111945:17)
    at async start (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:113032:16)
╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ B to open a browser, D to open Devtools, S to turn on (experimental) sharing, L to turn on local mode, X to exit   │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```
and it works well when I turn on local mode, but not without this.

Also, how do you think about showing log more clear way? For example, it has been showing logs like above cause it cannot catch exception on `fetch` process...

```
const response = await fetch(`${CF_API_BASE_URL}${resource}${queryString}`, {
    method: "GET",
    ...init,
    headers,
  });
```

How about wrapping this part as
```
try {
const response = await fetch(`${CF_API_BASE_URL}${resource}${queryString}`, {
    method: "GET",
    ...init,
    headers,
  });
} catch (e) {
    throw new Error(
      `bla bla bla -> ${e}`
    );
  }
```
?

**Comment by petebacondarwin** [comment]
What exact version of wrangler are you using? The latest beta release (0.0.12) should contain a change that shows better error messages for 403 Forbidden responses.

**Comment by djKooks** [comment]
@petebacondarwin I'm working with the newest code base (npm link package/wrangler). Currently it works as
```
$ wrangler dev index.js
⬣ Listening at http://localhost:8787
remote worker: Error: Failed to fetch /accounts/0d03575700684a90f4af1b90e87be426/workers/subdomain/edge-preview - 
10063: workers.api.error.subdomain_required
    at throwFetchError (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:109771:17)
    at fetchResult (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:109742:5)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async sessionToken (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:112138:28)
    at async previewToken (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:112156:53)
    at async createWorker (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:112180:17)
    at async start (/Users/kination/workspace/oss/wrangler2/packages/wrangler/wrangler-dist/cli.js:113346:16) {
  code: 10063
}
Tried to open a new devtools window when a previous one was already open.
Tried to open a new devtools window when a previous one was already open.
Tried to open a new devtools window when a previous one was already open.
Tried to open a new devtools window when a previous one was already open.
Tried to open a new devtools window when a previous one was already open.
╭─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ B to open a browser, D to open Devtools, S to turn on (experimental) sharing, L to turn on local mode, X to exit                        │
╰─────────────────
```
Is this works as you've intend?

**Comment by threepointone** [comment]
Ok I see what's happening here. You need to create a workers.dev subdomain on the dashboard. Go to dash.cloudflare.com, choose Workers in the left sidebar, and you should see a UI to set your workers subdomain in the right sidebar. 

Some follow up notes for us - 
- wrangler 1 had a command to set this subdomain from the command line. I dunno whether we should have it on Wrangler 2.
- I do think we should make users register a subdomain when they make a Cloudflare account
- in any case, we should handle this error better in the cli, and have messaging that asks a user to create a subdomain for workers.dev 

I'm going to reopen this issue, and we'll edit the title to reflect the problem and assign to someone to fix it later. 

Thank you for helping us find this, appreciate your patience!

**Comment by threepointone** [comment]
![image](https://user-images.githubusercontent.com/18808/151363399-b3e50001-bb11-4158-ba34-61d7188b27f0.jpeg)

**Comment by djKooks** [comment]
@threepointone thanks for notice~ I'll check

**Comment by djKooks** [comment]
@threepointone @petebacondarwin now error log not showing. Thanks for progress~
It shows as
```
$ wrangler dev index.js
⬣ Listening at http://localhost:8787
╭──────────────────────────────────────────────────────────────────────────────╮
│ B to open a browser, D to open Devtools, S to turn on (experimental)         │
│ sharing, L to turn on local mode, X to exit                                  │
╰──────────────────────────────────────────────────────────────────────────────╯
```

but in this case, is this local mode? (because I can check in localhost)
If it is seems it should not show `L to turn on local mode` 🤔

**Comment by threepointone** [comment]
This is as expected, we setup a proxy locally to access via `localhost`, but the actual worker is running on the edge. You can hit L to run the worker locally as well.

**Comment by djKooks** [comment]
@threepointone 
> actual worker is running on the edge
-> is this means I can see same page as localhost in worker `xxx.workers.dev` in non-local mode? Because I cannot see it...(localhost works well)

**Comment by petebacondarwin** [comment]
If I understand correct, when you use `wrangler dev` in remote mode, it is actually creating a temporary "preview" worker, not publishing to your `xxx.workers.dev` domain.

**Comment by petebacondarwin** [comment]
Closing as Wrangler appears to be working as expected, and this issue is turning into a mix of different resolved problems.
@djKooks  - Please open a new issue with any new problems you encounter.
